libr/core/cmd_debug.c

```diff
@@ -792,6 +796,224 @@ static void cmd_debug_modules(RCore *core, int mode) { // "dmm"
+			PRINTF_GREEN_ARENA (" 0x%"PFMT64x"->fd = ", tmp);
+			PRINTF_BLUE_ARENA ("0x%"PFMT64x, (ut64)(size_t)main_arena->bins[i]);
+			PRINT_GREEN_ARENA (", ");
+			PRINTF_GREEN_ARENA (" \t0x%"PFMT64x"->bk = ", tmp);
+			PRINTF_BLUE_ARENA ("0x%"PFMT64x, (ut64)(size_t)main_arena->bins[i + 1]);
+			PRINT_GREEN_ARENA (", ");
+		}
+	}
+	PRINT_GREEN_ARENA ("\n\t}\t\n");
+	PRINT_GREEN_ARENA ("\tbinmap = {");
+
+	for(i = 0; i < BINMAPSIZE; i++) {
+		PRINTF_BLUE_ARENA ("0x%x", (int)main_arena->binmap[i]);
+		if (i < 3) {
+			PRINT_GREEN_ARENA (",");
+		}
+	}
+	PRINT_GREEN_ARENA ("}\n");
+	PRINT_GREEN_ARENA ("\tnext = ");
+	PRINTF_BLUE_ARENA (" 0x%"PFMT64x, (ut64)(size_t)main_arena->next);
+	PRINT_GREEN_ARENA (",\n");
+	PRINT_GREEN_ARENA ("\tnext_free = ");
+	PRINTF_BLUE_ARENA (" 0x%"PFMT64x, (ut64)(size_t)main_arena->next_free);
+	PRINT_GREEN_ARENA (",\n");
+	PRINT_GREEN_ARENA ("\tsystem_mem = ");
+	PRINTF_BLUE_ARENA (" 0x%"PFMT64x, (ut64)(size_t)main_arena->system_mem);
+	PRINT_GREEN_ARENA (",\n");
+	PRINT_GREEN_ARENA ("\tmax_system_mem = ");
+	PRINTF_BLUE_ARENA (" 0x%"PFMT64x, (ut64)(size_t)main_arena->max_system_mem);
+	PRINT_GREEN_ARENA (",\n");
+	PRINT_GREEN_ARENA ("}\n\n");
+}
+
+static ut64 get_vaddr_symbol(const char *path, const char *symname) {
+	RListIter *iter;
+	RBinSymbol *s;
+	RCore *core = r_core_new ();
+	RList * syms = NULL;
+	ut64 vaddr = 0LL;
+
+	if (!core) return UT64_MAX;
+	r_bin_load (core->bin, path, 0, 0, 0, -1, false);
+	syms = r_bin_get_symbols (core->bin);
+	if (!syms) return UT64_MAX;
+	r_list_foreach (syms, iter, s) {
+		if (strstr (s->name, symname)) {
+			vaddr = s->vaddr;
+			break;
+		}
+	}
+	r_core_free (core);
+	return vaddr;
+}
+
+static void get_hash_debug_directory(const char *path, char *hash) {
+	RListIter *iter;
+	RBinSection *s;
+	RCore *core = r_core_new ();
+	RList * sects = NULL;
+	char buf[20] = {0};
+	int err, i, j = 0;
+
+	if (!core) return;
+	r_bin_load (core->bin, path, 0, 0, 0, -1, false);
+	sects = r_bin_get_sections (core->bin);
+	if (!sects) goto out_error;
+	r_list_foreach (sects, iter, s) {
+		if (strstr (s->name, ".note.gnu.build-id")) {
+			err = r_io_read_at (core->io, s->vaddr + 16, (ut8 *) buf, 20);
+			if (!err) {
+				eprintf ("We couldn't read from memory\n");
+				goto out_error; 
+			}
+			break;
+		}
+	}
+	for (i = 0; i < 20; i++) {
+		if (i <= 1) {
+			hash[i + 2 * j++] = (ut8) '/';
+		}
+		snprintf (hash + j + 2 * i, sizeof (hash), "%02x", (ut8) buf[i]);
+	}
+	strcat (hash, ".debug");
+out_error:
+	r_core_free (core);
+}
+
+static int cmd_debug_map_heap(RCore *core, const char *input) {
+	const char* help_msg[] = {
+		"Usage:", "dmh", " # Memory map heap",
+		"dmha", "", "Struct Malloc State (main_arena)",
+		"dmh?", "", "Show map heap help",
+		NULL
+	};
+
+	RListIter *iter;
+	RDebugMap *map;
+	RHeap_MallocState *main_arena;
+	static ut64 m_arena = UT64_MAX;
+
+	switch (input[0]) {
+	case 'a': // "dmha"
+		{
+		if (m_arena == UT64_MAX) {
```

Comment:
move into separate function? avoid that nested stackframe and make switch cases body smaller